### PR TITLE
Implement cert change handling

### PR DESCRIPTION
**Pull request by MykolaSuperman**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -17,6 +17,9 @@ option(WITH_TEST "build with test" OFF)
 option(WITH_COVERAGE "build with coverage" OFF)
 option(WITH_DOC "build with documenation" OFF)
 option(WITH_VCHAN "build with vchan" ON)
+option(WITH_IAM_API "build core api iam lib" ON)
```
Comment (al1img) [inline CMakeLists.txt]: Set external project option: It should not be option but rather we should set option for submodule. See SM external project options:
https://github.com/aosedge/aos_core_sm_cpp/blob/775fb6bdc302ef9398c5be73f2a7324e27e3c044/CMakeLists.txt#L29

Comment (al1img) [inline CMakeLists.txt]: Move to external project options: https://github.com/aosedge/aos_core_sm_cpp/blob/775fb6bdc302ef9398c5be73f2a7324e27e3c044/CMakeLists.txt#L29

File: src/cmclient/cmclient.cpp [hunk 2]
```diff
@@ -69,19 +71,59 @@ void CMClient::OnDisconnected()
...
+
+            mCMConnected = false;
+        }
+
+        while (!mShutdown) {
+            auto [credentials, err] = CreateCredentials();
+            if (!err.IsNone()) {
+                std::unique_lock lock {mMutex};
+
+                LOG_ERR() << "Failed to create credentials: error=" << err;
```
Comment (al1img) [inline src/cmclient/cmclient.cpp:94]: Empty line after log

File: src/cmclient/cmclient.cpp [hunk 3]
```diff
@@ -69,19 +71,59 @@ void CMClient::OnDisconnected()
...
+                mCV.wait_for(lock, cReconnectTimeout, [this] { return mShutdown.load(); });
+
+                continue;
+            }
+
+            mCredentials = credentials;
```
Comment (al1img) [inline src/cmclient/cmclient.cpp:103]: Looks like `mCredentials` is not guarded with mutex and is accessible from different threads.

File: src/communication/securechannel.cpp [hunk 4]
```diff
@@ -40,64 +39,97 @@ SecureChannel::SecureChannel(const config::Config& cfg, CommChannelItf& channel,
...
         SSL_free(mSsl);
+        mSsl = nullptr;
+    }
+
+    if (mCtx != nullptr) {
+        SSL_CTX_free(mCtx);
+        mCtx = nullptr;
+    }
+
+    mBioMethod.reset(nullptr);
```
Comment (al1img) [inline src/communication/securechannel.cpp:72]: It is necessary to do reset nullptr first?

File: src/communication/securechannel.hpp [hunk 5]
```diff
@@ -82,28 +85,30 @@ class SecureChannel : public CommChannelItf {
...
+    CommChannelItf*                     mChannel {};
+    common::iamclient::CertProviderItf* mCertProvider {};
+    crypto::CertLoaderItf*              mCertLoader {};
+    crypto::x509::ProviderItf*          mCryptoProvider {};
+    const config::Config*               mCfg {};
+    int                                 mPort {};
+    std::string                         mCertStorage;
+
+    SSL_CTX*                                              mCtx {};
+    SSL*                                                  mSsl {};
```
Comment (al1img) [inline src/communication/securechannel.hpp]: rename to `mSSL`

File: src/communication/securechannel.hpp [hunk 6]
```diff
@@ -82,28 +85,30 @@ class SecureChannel : public CommChannelItf {
...
+    std::unique_ptr<BIO_METHOD, decltype(&BIO_meth_free)> mBioMethod {nullptr, BIO_meth_free};
+    std::atomic<bool>                                     mConnected {};
 };
 
 } // namespace aos::mp::communication
 
-#endif /* SECURECHANNEL_HPP_ */
+#endif /* SECURECHANNEL_HPP_ */
```
Comment (al1img) [inline src/communication/securechannel.hpp]: new line at the end

File: .github/workflows/build_test.yaml [hunk 7]
```diff
@@ -38,7 +38,15 @@ jobs:
             -DWITH_CORELIB=ON -DWITH_TEST=ON -DWITH_VCHAN=OFF -DWITH_COVERAGE=ON -DCMAKE_BUILD_TYPE=Debug
 
           cd build
-          make; make tests; make coverage
+          make; make tests;   make coverage
```
Comment (al1img) [inline .github/workflows/build_test.yaml]: revert this line

File: .gitmodules [hunk 8]
```diff
@@ -1,3 +1,3 @@
 [submodule "external/aos_core_common_cpp"]
 	path = external/aos_core_common_cpp
-	url = https://github.com/aosedge/aos_core_common_cpp.git
```
Comment (al1img) [inline .gitmodules:3]: Remove this commit

File: src/iamclient/publicnodeclient.cpp [hunk 9]
```diff
@@ -62,6 +64,51 @@ void PublicNodeClient::OnDisconnected()
     Close();
 }
 
+void PublicNodeClient::OnCertChanged(const iamanager::v5::CertInfo& info)
+{
+    LOG_DBG() << "Certificate changed: info=" << info.DebugString().c_str();
+
+    std::thread([this] {
```
Comment (mlohvynenko) [inline src/iamclient/publicnodeclient.cpp]: consider using poco thread pool

File: CMakeLists.txt (hunk 1 above)
Comment (MykolaSuperman) [reply CMakeLists.txt]: Done

Comment (MykolaSuperman) [reply CMakeLists.txt]: Done

File: src/cmclient/cmclient.cpp (hunk 2 above)
Comment (MykolaSuperman) [reply src/cmclient/cmclient.cpp:94]: Done

File: src/cmclient/cmclient.cpp (hunk 3 above)
Comment (MykolaSuperman) [reply src/cmclient/cmclient.cpp:103]: Done

File: src/communication/securechannel.cpp (hunk 4 above)
Comment (MykolaSuperman) [reply src/communication/securechannel.cpp:72]: Done

File: src/communication/securechannel.hpp (hunk 5 above)
Comment (MykolaSuperman) [reply src/communication/securechannel.hpp]: Done

File: src/communication/securechannel.hpp (hunk 6 above)
Comment (MykolaSuperman) [reply src/communication/securechannel.hpp]: Done

File: .gitmodules (hunk 8 above)
Comment (MykolaSuperman) [reply .gitmodules:3]: Done

File: src/iamclient/publicnodeclient.cpp (hunk 9 above)
Comment (MykolaSuperman) [reply src/iamclient/publicnodeclient.cpp]: Done

**Review by mlohvynenko — APPROVED** [review]

File: src/cmclient/cmclient.cpp [hunk 10]
```diff
@@ -69,19 +70,64 @@ void CMClient::OnDisconnected()
...
+
+            std::lock_guard lock {mMutex};
+
+            mCredentials = credentials;
+
+            return;
+        }
+    };
+
+    mThreadPool.start(*makeRunnable(std::move(task)));
```
Comment (mlohvynenko) [inline src/cmclient/cmclient.cpp]: start might throw
Add exception handling

File: src/iamclient/publicnodeclient.cpp [hunk 11]
```diff
@@ -62,6 +64,54 @@ void PublicNodeClient::OnDisconnected()
...
+                return;
+            }
+
+            mCredentialList.back() = res.mValue;
+
+            return;
+        }
+    };
+
+    mThreadPool.start(*makeRunnable(std::move(task)));
```
Comment (mlohvynenko) [inline src/iamclient/publicnodeclient.cpp]: start might throw
Add exception handling

File: src/cmclient/cmclient.cpp (hunk 10 above)
Comment (MykolaSuperman) [reply src/cmclient/cmclient.cpp]: Done

File: src/iamclient/publicnodeclient.cpp (hunk 11 above)
Comment (MykolaSuperman) [reply src/iamclient/publicnodeclient.cpp]: Done

File: CMakeLists.txt [hunk 12]
```diff
@@ -17,6 +17,7 @@ option(WITH_TEST "build with test" OFF)
 option(WITH_COVERAGE "build with coverage" OFF)
 option(WITH_DOC "build with documenation" OFF)
 option(WITH_VCHAN "build with vchan" ON)
+option(WITH_AOS_PROTOCOL "build aos protocol lib" ON)
```
Comment (al1img) [inline CMakeLists.txt]: Is it still required?

File: src/cmclient/cmclient.cpp [hunk 13]
```diff
@@ -69,19 +71,70 @@ void CMClient::OnDisconnected()
...
+            mCredentials = credentials;
+
+            return;
+        }
+    };
+
+    try {
+        mThreadPool.start(*makeRunnable(std::move(task)));
+    } catch (const Poco::Exception& e) {
+        LOG_ERR() << "Failed to start cert change task: " << e.displayText().c_str();
```
Comment (al1img) [inline src/cmclient/cmclient.cpp]: error=

File: src/cmclient/cmclient.cpp [hunk 14]
```diff
@@ -69,19 +71,70 @@ void CMClient::OnDisconnected()
...
+    } catch (const std::exception& e) {
+        LOG_ERR() << "Failed to start cert change task: " << e.what();
```
Comment (al1img) [inline src/cmclient/cmclient.cpp]: ditto

File: src/iamclient/publicnodeclient.cpp [hunk 15]
```diff
@@ -62,6 +65,60 @@ void PublicNodeClient::OnDisconnected()
...
+            mCredentialList.back() = res.mValue;
+
+            return;
+        }
+    };
+
+    try {
+        mThreadPool.start(*makeRunnable(std::move(task)));
+    } catch (const Poco::Exception& e) {
+        LOG_ERR() << "Failed to start cert change task: " << e.displayText().c_str();
```
Comment (al1img) [inline src/iamclient/publicnodeclient.cpp]: error=

File: src/iamclient/publicnodeclient.cpp [hunk 16]
```diff
@@ -62,6 +65,60 @@ void PublicNodeClient::OnDisconnected()
...
+    } catch (const std::exception& e) {
+        LOG_ERR() << "Failed to start cert change task: " << e.what();
```
Comment (al1img) [inline src/iamclient/publicnodeclient.cpp]: ditto

File: CMakeLists.txt (hunk 12 above)
Comment (MykolaSuperman) [reply CMakeLists.txt]: no, forgot to remove

File: src/cmclient/cmclient.cpp (hunk 13 above)
Comment (MykolaSuperman) [reply src/cmclient/cmclient.cpp]: Done

File: src/cmclient/cmclient.cpp (hunk 14 above)
Comment (MykolaSuperman) [reply src/cmclient/cmclient.cpp]: Done

File: src/iamclient/publicnodeclient.cpp (hunk 15 above)
Comment (MykolaSuperman) [reply src/iamclient/publicnodeclient.cpp]: Done

File: src/iamclient/publicnodeclient.cpp (hunk 16 above)
Comment (MykolaSuperman) [reply src/iamclient/publicnodeclient.cpp]: Done

**Review by al1img — APPROVED** [review]

**Pull request merged**